prov/util/src/util_mem_monitor.c

```diff
@@ -259,12 +260,26 @@ void ofi_monitor_notify(struct ofi_mem_monitor *monitor,
 {
 	struct ofi_mr_cache *cache;
 
-	dlist_foreach_container(&monitor->list, struct ofi_mr_cache,
+	dlist_ts_foreach_container(&monitor->list, struct ofi_mr_cache,
 				cache, notify_entries[monitor->iface]) {
 		ofi_mr_cache_notify(cache, addr, len);
 	}
 }
 
+/* Must be called holding monitor lock */
+void ofi_monitor_flush(struct ofi_mem_monitor *monitor)
+{
+	struct ofi_mr_cache *cache;
+
+	dlist_ts_foreach_container(&monitor->list, struct ofi_mr_cache,
```

Comment:
This must be paired with dlist_tx_foreach_end(), so that the dlist lock is released.

But this is basically moving the problem of holding the mm_lock to the dlist lock.  If we hold the dlist lock during flush, and that frees memory, the uffd thread will run, try to call notify, which will hang trying to get the lock.

Okay, how did the original submission not have these locking issues?  Or did we overlook a locking problem before?

Reply:
I'm pretty sure we overlooked a locking problem before. See my other comment.